core/pouch/migrations.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * @module core/pouch/migrations
+ * @flow
+ */
+
+const { PouchError } = require('.')
+
+/*::
+import type { Pouch } from './'
+import type { Metadata } from '../metadata'
+
+type SchemaVersion = number
+
+type MigrationData = {
+  migrated: string[],
+  errors: PouchError[]
+}
+type MigrationNoop = MigrationData & {
+  type: 'MigrationNoop'
+}
+type MigrationComplete = MigrationData & {
+  type: 'MigrationComplete',
+}
+type MigrationFailed = MigrationData & {
+  type: 'MigrationFailed',
+}
+type MigrationResult = MigrationNoop|MigrationComplete|MigrationFailed
+
+export type Migration = {
+  baseSchemaVersion: SchemaVersion,
+  targetSchemaVersion: SchemaVersion,
+  description: string,
+  affectedDocs: (Metadata[]) => Metadata[],
+  run: (Metadata[]) => Metadata[]
+}
+*/
+
+const SCHEMA_DOC_ID = '_local/schema'
+const SCHEMA_INITIAL_VERSION = 0
+const INITIAL_SCHEMA = {
+  _id: SCHEMA_DOC_ID,
+  version: SCHEMA_INITIAL_VERSION
+}
+const MIGRATION_RESULT_NOOP = 'MigrationNoop'
+const MIGRATION_RESULT_COMPLETE = 'MigrationComplete'
+const MIGRATION_RESULT_FAILED = 'MigrationFailed'
+
+const migrations /*: Migration[] */ = []
+
+function MIGRATION_NOOP() {
+  return { type: MIGRATION_RESULT_NOOP, errors: [], migrated: [] }
+}
+
+async function currentSchemaVersion(
+  pouch /*: Pouch */
+) /*: Promise<SchemaVersion> */ {
+  const schema = (await pouch.byIdMaybeAsync(SCHEMA_DOC_ID)) || INITIAL_SCHEMA
+  return schema.version || SCHEMA_INITIAL_VERSION
+}
+
+async function updateSchemaVersion(
+  version /*: SchemaVersion */,
+  pouch /*: Pouch */
+) {
+  const schema = (await pouch.byIdMaybeAsync(SCHEMA_DOC_ID)) || INITIAL_SCHEMA
+  return pouch.db.put({ ...schema, version })
+}
+
+async function migrate(
+  migration /*: Migration */,
+  pouch /*: Pouch */
+) /*: Promise<MigrationResult> */ {
+  if ((await currentSchemaVersion(pouch)) !== migration.baseSchemaVersion) {
+    return new MIGRATION_NOOP()
+  } else {
+    const docs = await pouch.byRecursivePathAsync('')
+    const affectedDocs = migration.affectedDocs(docs)
+    const migratedDocs = migration.run(affectedDocs)
+    const result = await save(migratedDocs, pouch)
+    if (result.type !== MIGRATION_RESULT_FAILED) {
+      await updateSchemaVersion(migration.targetSchemaVersion, pouch)
+    }
+    return result
+  }
+}
+
+async function save(
+  docs /*: Metadata[] */,
+  pouch /*: Pouch */
+) /*: Promise<MigrationResult> */ {
+  const migrationResult /* MigrationResult */ = new MIGRATION_NOOP()
```

Comment:
It feels weird to use `new` with something declared as a function, and not as a class

Reply:
It's idiomatic javascript though. I'm only using classes for very large classes or when Flow doesn't let me use a function as a type.

Reply:
Sorry I misunderstood your question. I've renamed the function to `migrationNoop` and will use it as a simple function.